client/question/domain/CodeEvalResultObjectFactorySpec.js

```diff
@@ -165,6 +165,132 @@ describe('CodeEvalResultObjectFactory', function() {
+          performanceTests: []
+        }),
+        TaskObjectFactory.create({
+          instructions: [''],
+          prerequisiteSkills: [''],
+          acquiredSkills: [''],
+          inputFunctionName: null,
+          outputFunctionName: null,
+          mainFunctionName: 'mockMainFunction',
+          languageSpecificTips: {
+            python: []
+          },
+          testSuites: [{
+            id: 'GENERAL_CASE',
+            humanReadableName: 'the general case',
+            testCases: [{
+              input: 'task_2_correctness_test_1',
+              allowedOutputs: [true]
+            }, {
+              input: 'task_2_correctness_test_2',
+              allowedOutputs: [true]
+            }]
+          }],
+          buggyOutputTests: [],
+          suiteLevelTests: [],
+          performanceTests: []
+        })
+      ];
+
+      var codeEvalResult1 = CodeEvalResultObjectFactory.create(
+        CODE, OUTPUT, [[[true, true]], [[true, true]]],
+        BUGGY_OUTPUT_TEST_RESULTS, PERFORMANCE_TEST_RESULTS, ERROR_STRING,
+        ERROR_INPUT);
+      expect(codeEvalResult1.getIndexOfFirstFailedTest(tasks))
+        .toEqual(3);
+
+      var codeEvalResult2 = CodeEvalResultObjectFactory.create(
+        CODE, OUTPUT, [[[true, false]], [[true, true]]],
+        BUGGY_OUTPUT_TEST_RESULTS, PERFORMANCE_TEST_RESULTS, ERROR_STRING,
+        ERROR_INPUT);
+      expect(codeEvalResult2.getIndexOfFirstFailedTest(tasks))
+        .toEqual(1);
+
+      var codeEvalResult3 = CodeEvalResultObjectFactory.create(
+        CODE, OUTPUT, [[[true, true]], [[false, true]]],
+        BUGGY_OUTPUT_TEST_RESULTS, PERFORMANCE_TEST_RESULTS, ERROR_STRING,
+        ERROR_INPUT);
+      expect(codeEvalResult3.getIndexOfFirstFailedTest(tasks))
+        .toEqual(2);
+
+      var codeEvalResult4 = CodeEvalResultObjectFactory.create(
+        CODE, OUTPUT, [[[true, false]], [[false, true]]],
+        BUGGY_OUTPUT_TEST_RESULTS, PERFORMANCE_TEST_RESULTS, ERROR_STRING,
+        ERROR_INPUT);
+      expect(codeEvalResult4.getIndexOfFirstFailedTest(tasks))
+        .toEqual(1);
+
+      var codeEvalResult5 = CodeEvalResultObjectFactory.create(
+        CODE, OUTPUT, [], BUGGY_OUTPUT_TEST_RESULTS, PERFORMANCE_TEST_RESULTS,
+        ERROR_STRING, ERROR_INPUT);
+      expect(codeEvalResult5.getIndexOfFirstFailedTest(tasks))
+        .toEqual(0);
+
+      var codeEvalResult6 = CodeEvalResultObjectFactory.create(
+        CODE, OUTPUT, [[[false, false]], [[false, true]]],
+        BUGGY_OUTPUT_TEST_RESULTS, PERFORMANCE_TEST_RESULTS,
+        ERROR_STRING, ERROR_INPUT);
+      expect(codeEvalResult6.getIndexOfFirstFailedTest(tasks))
+        .toEqual(0);
+
+      var codeEvalResult7 = CodeEvalResultObjectFactory.create(
+        CODE, OUTPUT, [[[true, true]]],
+        BUGGY_OUTPUT_TEST_RESULTS, PERFORMANCE_TEST_RESULTS,
+        ERROR_STRING, ERROR_INPUT);
+      expect(codeEvalResult7.getIndexOfFirstFailedTest(tasks))
+        .toEqual(1);
```

Comment:
This seems wrong? Shouldn't it be 2?

Also, one question -- is it possible for this._observedOutputs to have a different length compared to the length of `tasks` (or for its subcomponents to have lengths not matching the corresponding subparts of `tasks`)? If so then perhaps this should be more explicitly handled in the main function.

Reply:
I'll modify this to null later to signify that all test cases were passed, but originally it was 1 since the 2nd test case (1 via 0-indexing) is the last test case to be executed. 

If a user only finished 1/3 tasks for example, `this._observedOutputs` will be shorter than the length of `tasks` since `tasks` will have all of those for that question, which was tested here since it gets handled in the `getIndexOfFirstFailedTest` function (loops over `this._observedOutputs` rather than `tasks`). Would it be better to test this elsewhere?

Reply:
But they passed the 2nd test case (the one that was 0-indexed as 1), right? So it can't be the "first failed test case". If anything, perhaps it should be the first case for which we got no results?

I think the right thing to do depends on the stdout behaviour we want when a task is completely passed (and in other cases with partial results). Did you get a decision on that? It might be worth enumerating them.

Reply:
After my conversation with John, it seems that the best option is to display whichever test case was last run (regardless of if the code completely passed or not). In this case, would it be best to modify the function name to more accurately depict that it returns the last test case executed?